### ISSUE TITLE
Parameter gradients not zeroed-out in gradient descent

### ISSUE THREAD

**Issue by kawu** [kawu/hhu-dl-materials-2020#11]
This bug was introduced in the section on PyTorch [optimizers](https://github.com/kawu/hhu-dl-materials-2020/tree/main/gradient-descent#optimisers), in particular in this piece of code:
```python
# Perform SGD for 1000 epochs
for k in range(1000):
    # Optional: use random dataset permutation in each epoch
    for i in torch.randperm(len(enc_data)):
        x, y = enc_data[i]
        loss(baseline(x), y).backward()
        optim.step()	# version of `nudge` provided by `Adam`
```
In contrast with [nudge](https://github.com/kawu/hhu-dl-materials-2020/tree/main/gradient-descent#backward-calculation) defined earlier:
```python
def nudge(model: nn.Module, learning_rate=0.1):
    '''Nudge the parameters of the `model` along its gradient'''
    # Switch off the gradient calculation machinery
    with torch.no_grad():
        for param in model.parameters():
            param -= learning_rate * param.grad  # type: ignore
        # The `backward` method is cumulative, so we have to reset
        # to zero all the gradients
        model.zero_grad()
```
`optim.step()` **does not zero-out the gradients** after updating the parameters.

One effect of this issue is that the learning rate has be set to a very low value to make training stable.
